network/p2p/acp118/aggregator.go

```diff
@@ -0,0 +1,260 @@
+		if signerBitSet.Contains(i) {
+			aggregatedStakeWeight.Add(aggregatedStakeWeight, new(big.Int).SetUint64(v.Weight))
+			continue
+		}
+
+		publicKeysToValidators[v.PublicKey] = indexedValidator{
+			Index:     i,
+			Validator: v,
+		}
+
+		for _, nodeID := range v.NodeIDs {
+			numRequests += 1
+			nodeIDsToPublicKeys[nodeID] = v.PublicKey
+		}
+
+		nonSigners = append(nonSigners, v.PublicKey)
+	}
+
+	// Account for requested signatures + the signature that was provided
+	signatures := make([]*bls.Signature, 0, len(nonSigners)+1)
+	if bitSetSignature.Signature != [bls.SignatureLen]byte{} {
+		blsSignature, err := bls.SignatureFromBytes(bitSetSignature.Signature[:])
+		if err != nil {
+			return nil, nil, nil, false, fmt.Errorf("failed to parse bls signature: %w", err)
+		}
+		signatures = append(signatures, blsSignature)
+	}
+
+	results := make(chan result)
+	handler := responseHandler{
+		message:                message,
+		publicKeysToValidators: publicKeysToValidators,
+		nodeIDsToPublicKeys:    nodeIDsToPublicKeys,
+		results:                results,
+	}
+
+	for _, pk := range nonSigners {
+		validator := publicKeysToValidators[pk]
+		for _, nodeID := range validator.NodeIDs {
+			if err := s.client.AppRequest(ctx, set.Of(nodeID), requestBytes, handler.HandleResponse); err != nil {
+				return nil, nil, nil, false, fmt.Errorf("failed to send aggregation request: %w", err)
+			}
+		}
+	}
+
+	minThreshold := new(big.Int).Mul(totalStakeWeight, new(big.Int).SetUint64(quorumNum))
+	minThreshold.Div(minThreshold, new(big.Int).SetUint64(quorumDen))
+
+	for i := 0; i < numRequests; i++ {
+		select {
+		case <-ctx.Done():
+			// Try to return whatever progress we have if the context is cancelled
+			msg, err := newWarpMessage(message, signerBitSet, signatures)
+			if err != nil {
+				return nil, nil, nil, false, err
+			}
+
+			return msg, aggregatedStakeWeight, totalStakeWeight, false, nil
+		case result := <-results:
+			if result.Err != nil {
+				s.log.Debug(
+					"dropping response",
+					zap.Stringer("nodeID", result.NodeID),
+					zap.Error(err),
+				)
+				continue
+			}
+
+			if signerBitSet.Contains(result.Validator.Index) {
+				s.log.Debug(
+					"dropping duplicate signature",
+					zap.Stringer("nodeID", result.NodeID),
+					zap.Error(err),
+				)
+				continue
+			}
+
+			signatures = append(signatures, result.Signature)
+			signerBitSet.Add(result.Validator.Index)
+			aggregatedStakeWeight.Add(aggregatedStakeWeight, new(big.Int).SetUint64(result.Validator.Weight))
+
+			if aggregatedStakeWeight.Cmp(minThreshold) != -1 {
+				msg, err := newWarpMessage(message, signerBitSet, signatures)
+				if err != nil {
+					return nil, nil, nil, false, err
+				}
+
+				return msg, aggregatedStakeWeight, totalStakeWeight, true, nil
+			}
+		}
+	}
+
+	msg, err := newWarpMessage(message, signerBitSet, signatures)
+	if err != nil {
+		return nil, nil, nil, false, err
+	}
+
+	return msg, aggregatedStakeWeight, totalStakeWeight, true, nil
+}
```

Comment:
(optional nit) could we add a comment to explain the success/failure cases here? Reading through this, I have to pause to see why we handle each signature in the loop, but may exit the loop without having terminated.

I think I'm just used to seeing the previous style and what iirc is used in the EVM of checking for both terminate on reached required threshold and can no longer reach required threshold.

iiuc the new logic is terminate in the loop if we reach the termination threshold, otherwise we will process responses until either the context expires or we've processed every response (max out signature weight up to ctx expiration).

Reply:
> iiuc the new logic is terminate in the loop if we reach the termination threshold, otherwise we will process responses until either the context expires or we've processed every response (max out signature weight up to ctx expiration).

Yes that's correct. I can add comments to make this more clear.